managed/services/telemetry/transform.go

```diff
@@ -0,0 +1,96 @@
+// Copyright (C) 2017 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+package telemetry
+
+import (
+	"encoding/json"
+	pmmv1 "github.com/percona-platform/saas/gen/telemetry/events/pmm"
+	"github.com/pkg/errors"
+)
+
+type itemsType []map[string]any
+
+func transformToJSON(config *Config, metrics []*pmmv1.ServerMetric_Metric) ([]*pmmv1.ServerMetric_Metric, error) {
+	if metrics == nil {
+		return nil, nil
+	}
+
+	if config.Transform == nil {
+		return nil, errors.Errorf("no transformation config is set")
+	}
+
+	if config.Transform.Type != JSONTransformType {
+		return nil, errors.Errorf("not supported transformation type [%s], it must be [%s]", config.Transform.Type, JSONTransformType)
+	}
+
+	if len(config.Data) == 0 || config.Data[0].MetricName == "" {
+		return nil, errors.Errorf("invalid metrics config")
+	}
+
+	firstMetric := config.Data[0].MetricName
+
+	result := map[string]itemsType{}
+
+	var items []map[string]any
+	var next map[string]any
+	for _, metric := range metrics {
+		isFirstMetric := metric.Key == firstMetric
```

Comment:
Do you think we need this check now? As I understand now we set metric in datasource's implementations even if metric has empty value.

Reply:
Oh, now I got why do we need that.

Reply:
added comment to make it a bit more clear